File: code/packages/elixir/type_checker_protocol/lib/coding_adventures/type_checker_protocol.ex
defmodule CodingAdventures.TypeCheckerProtocol.TypeErrorDiagnostic do
  @enforce_keys [:message, :line, :column]
  defstruct [:message, :line, :column]

  @type t :: %__MODULE__{
          message: String.t(),
          line: pos_integer(),
          column: pos_integer()
        }
end

defmodule CodingAdventures.TypeCheckerProtocol.TypeCheckResult do
  @enforce_keys [:typed_ast, :errors, :ok]
  defstruct [:typed_ast, :errors, :ok]

  @type t :: %__MODULE__{
          typed_ast: term(),
          errors: [CodingAdventures.TypeCheckerProtocol.TypeErrorDiagnostic.t()],
          ok: boolean()
        }
end

defmodule CodingAdventures.TypeCheckerProtocol do
  alias CodingAdventures.TypeCheckerProtocol.TypeCheckResult
  alias CodingAdventures.TypeCheckerProtocol.TypeErrorDiagnostic

  @callback check(term()) :: TypeCheckResult.t()

  def new_result(typed_ast, errors \\ []) do
    %TypeCheckResult{typed_ast: typed_ast, errors: errors, ok: errors == []}
  end

  def new_diagnostic(message, line \\ 1, column \\ 1) do
    %TypeErrorDiagnostic{message: message, line: line, column: column}
  end
end

defmodule CodingAdventures.TypeCheckerProtocol.GenericTypeChecker do
  alias CodingAdventures.TypeCheckerProtocol

  defstruct hooks: %{}, errors: [], node_kind: nil, locate: nil

  def new(opts \\ []) do
    %__MODULE__{
      hooks: %{},
      errors: [],
      node_kind: Keyword.get(opts, :node_kind),
      locate: Keyword.get(opts, :locate, fn _ -> {1, 1} end)
    }
  end

  def reset(%__MODULE__{} = checker) do
    %{checker | errors: []}
  end

  def register_hook(%__MODULE__{} = checker, phase, kind, hook) do
    key_kind = if kind == "*", do: "*", else: normalize_kind(kind)
    key = "#{phase}:#{key_kind}"
    hooks = Map.update(checker.hooks, key, [hook], fn existing -> existing ++ [hook] end)
    %{checker | hooks: hooks}
  end

  def dispatch(%__MODULE__{} = checker, phase, node, args \\ []) do
    kind =
      if is_function(checker.node_kind, 1) do
        checker.node_kind.(node) |> to_string() |> normalize_kind()
      else
        ""
      end

    Enum.reduce_while(["#{phase}:#{kind}", "#{phase}:*"], nil, fn key, _acc ->
      hooks = Map.get(checker.hooks, key, [])

      result =
        Enum.reduce_while(hooks, :not_handled, fn hook, _inner ->
          value = hook.(node, args)

          if value == :not_handled do
            {:cont, :not_handled}
          else
            {:halt, value}
          end
        end)

      if result == :not_handled do
        {:cont, nil}
      else
        {:halt, result}
      end
    end)
  end

  def not_handled, do: :not_handled

  def error(%__MODULE__{} = checker, message, subject) do
    {line, column} = checker.locate.(subject)
    diagnostic = TypeCheckerProtocol.new_diagnostic(message, line, column)
    %{checker | errors: checker.errors ++ [diagnostic]}
  end

  def check(%__MODULE__{} = checker, ast, run_fun) when is_function(run_fun, 2) do
    checker = reset(checker)
    checker = run_fun.(checker, ast)
    TypeCheckerProtocol.new_result(ast, checker.errors)
  end

  defp normalize_kind(kind) do
    kind
    |> String.graphemes()
    |> Enum.reduce({"", false}, fn grapheme, {acc, last_underscore} ->
      if grapheme =~ ~r/^[[:alnum:]]$/u do
        {acc <> grapheme, false}
      else
        if last_underscore do
          {acc, true}
        else
          {acc <> "_", true}
        end
      end
    end)
    |> elem(0)
    |> String.trim("_")
  end
end

File: code/packages/csharp/type-checker-protocol/TypeCheckerProtocol.cs
using System.Text.RegularExpressions;

namespace CodingAdventures.TypeCheckerProtocol;

/// <summary>
/// A single type-checking diagnostic with a source location.
/// </summary>
public sealed record TypeErrorDiagnostic(string Message, int Line, int Column);

/// <summary>
/// The result of a type-checking pass.
/// </summary>
public sealed class TypeCheckResult<TAst>
{
    /// <summary>Create a result with an AST and optional diagnostics.</summary>
    public TypeCheckResult(TAst typedAst, IEnumerable<TypeErrorDiagnostic>? errors = null)
    {
        TypedAst = typedAst;
        Errors = (errors ?? []).ToArray();
    }

    /// <summary>The typed or partially typed AST.</summary>
    public TAst TypedAst { get; }

    /// <summary>Diagnostics collected during checking.</summary>
    public IReadOnlyList<TypeErrorDiagnostic> Errors { get; }

    /// <summary>True when no diagnostics were reported.</summary>
    public bool Ok => Errors.Count == 0;
}

/// <summary>
/// Generic contract for language-specific type checkers.
/// </summary>
public interface ITypeChecker<TAstIn, TAstOut>
{
    /// <summary>Type-check an AST and return its typed form plus diagnostics.</summary>
    TypeCheckResult<TAstOut> Check(TAstIn ast);
}

/// <summary>
/// Marker value a hook may return to ask dispatch to keep searching.
/// </summary>
public sealed class NotHandled
{
    private NotHandled()
    {
    }

    /// <summary>The singleton marker value.</summary>
    public static NotHandled Value { get; } = new();
}

/// <summary>
/// Reusable base class for AST-driven type checkers.
/// </summary>
public abstract class GenericTypeChecker<TAst> : ITypeChecker<TAst, TAst>
{
    private readonly List<TypeErrorDiagnostic> _errors = [];
    private readonly Dictionary<(string Phase, string Kind), List<Func<TAst, object?[], object?>>> _hooks = [];

    /// <summary>Run the checker and return the AST plus collected diagnostics.</summary>
    public TypeCheckResult<TAst> Check(TAst ast)
    {
        _errors.Clear();
        Run(ast);
        return new TypeCheckResult<TAst>(ast, _errors);
    }

    /// <summary>Register a hook for a phase and normalized node kind.</summary>
    public void RegisterHook(string phase, string kind, Func<TAst, object?[], object?> hook)
    {
        ArgumentNullException.ThrowIfNull(phase);
        ArgumentNullException.ThrowIfNull(kind);
        ArgumentNullException.ThrowIfNull(hook);

        var key = (phase, NormalizeKind(kind));
        if (!_hooks.TryGetValue(key, out var hooks))
        {
            hooks = [];
            _hooks[key] = hooks;
        }

        hooks.Add(hook);
    }

    /// <summary>Dispatch a node to the first matching hook.</summary>
    public object? Dispatch(string phase, TAst node, object?[]? args = null, object? defaultValue = null)
    {
        ArgumentNullException.ThrowIfNull(phase);
        var normalized = NormalizeKind(NodeKind(node));
        foreach (var key in new[] { (phase, normalized), (phase, "*") })
        {
            if (!_hooks.TryGetValue(key, out var hooks))
            {
                continue;
            }

            foreach (var hook in hooks)
            {
                var result = hook(node, args ?? []);
                if (!ReferenceEquals(result, NotHandled.Value))
                {
                    return result;
                }
            }
        }

        return defaultValue;
    }

    /// <summary>Normalize a node kind for hook lookup.</summary>
    public static string NormalizeKind(string? kind)
    {
        if (string.IsNullOrEmpty(kind))
        {
            return string.Empty;
        }

        return Regex.Replace(kind, @"\W+", "_").Trim('_');
    }

    /// <summary>Run concrete type-checking logic.</summary>
    protected abstract void Run(TAst ast);

    /// <summary>Return a language-specific kind label for a node.</summary>
    protected abstract string? NodeKind(TAst node);

    /// <summary>Return the diagnostic location for a subject.</summary>
    protected virtual (int Line, int Column) Locate(object subject) => (1, 1);

    /// <summary>Append one diagnostic for a subject.</summary>
    protected void Error(string message, object subject)
    {
        var (line, column) = Locate(subject);
        _errors.Add(new TypeErrorDiagnostic(message, line, column));
    }
}

/// <summary>
/// Package metadata.
/// </summary>
public static class TypeCheckerProtocolPackage
{
    /// <summary>The package version.</summary>
    public const string Version = "0.1.0";
}
